typescript/sdk/src/core/HyperlaneRelayer.ts

```diff
@@ -31,19 +36,64 @@ const DerivedHookConfigWithAddressSchema =
 const DerivedIsmConfigWithAddressSchema =
   IsmConfigSchema.and(WithAddressSchema);
 
+const BacklogMessageSchema = z.object({
+  attempts: z.number(),
+  lastAttempt: z.number(),
+  message: z.string(),
+  dispatchTx: z.string(),
+});
+
+const MessageBacklogSchema = z.array(BacklogMessageSchema);
+
 export const RelayerCacheSchema = z.object({
   hook: z.record(z.record(DerivedHookConfigWithAddressSchema)),
   ism: z.record(z.record(DerivedIsmConfigWithAddressSchema)),
+  backlog: MessageBacklogSchema,
 });
 
 type RelayerCache = z.infer<typeof RelayerCacheSchema>;
 
+type MessageWhitelist = ChainMap<Set<Address>>;
+
+// message must have origin and destination chains in the whitelist
+// if whitelist has non-empty address set for chain, message must have sender and recipient in the set
+export function messageMatchesWhitelist(
+  whitelist: MessageWhitelist,
+  message: ParsedMessage,
+): boolean {
+  const originAddresses = whitelist[message.originChain ?? message.origin];
```

Comment:
iiuc, this is the chainId, and they are no longer unique. 

if there are chains with the same originId, but only 1 is whitelisted, does this mean they are all whitelisted?

Reply:
this is the domain ID as encoded by the mailbox onchain

Reply:
if you were referring to the `originChain` thats the human readable identifier which should be unique